### PR TITLE
8334599: Improve code from JDK-8302671

### PR DESCRIPTION
**Pull request by TheShermanTanker**
In [JDK\-8302671](https://bugs.openjdk.org/browse/JDK-8302671) I fixed a memmove decay bug by rewriting a sizeof on an array to an explicit size of 256, but this is a bit of a band aid fix. It's come to my attention that in C++, one can pass an array by reference, which causes sizeof to work correctly on an array and has the added bonus of enforcing an array of that size on the arguments passed to that method. I've reverted my change from 8302671 and instead explicitly made kstate an array reference so that sizeof works on the array as expected, and that the array size can be explicitly set in the array brackets

Verification: https://godbolt.org/z/Ezj76eWWY and GitHub Actions

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334599](https://bugs.openjdk.org/browse/JDK-8334599): Improve code from JDK-8302671 (**Enhancement** - P4)

### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) 🔄 Re-review required (review applies to [b98068e9](https://git.openjdk.org/jdk/pull/19798/files/b98068e958da7ba74a574bb5fea174f492a5fd32))
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19798/head:pull/19798` \
`$ git checkout pull/19798`

Update a local copy of the PR: \
`$ git checkout pull/19798` \
`$ git pull https://git.openjdk.org/jdk.git pull/19798/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19798`

View PR using the GUI difftool: \
`$ git pr show -t 19798`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19798.diff">https://git.openjdk.org/jdk/pull/19798.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19798#issuecomment-2180127636)

### CODE REVIEW

**Review by prrace — COMMENTED** [review]
Harmless, I hope, but it doesn't change that 256 is hard-coded in multiple places AND that the caller has to get it right too .. if the caller passes a shorter array things can still go wrong.
Why can't we EVERYWHERE use the definition used to create the array passed in
        KB_STATE_SIZE = 256
instead of a mixture.

**Comment by TheShermanTanker** [comment]
Yes, it doesn't change any existing behaviour, it's just a more robust way of expressing what was already there before, so it should be harmless. I'm not sure I follow your other concern about it being a mixture though, I checked and this method is static and only used in this file, and the few callsites that do use it all initialize their arrays to KB\_STATE\_SIZE. Do you mean getting rid of the hardcoded 256 in this method's array parameter? The other thing about passing an array by reference means if the caller doesn't get the array size exactly right, it will not result in a silent bug, it is an immediate compilation error, so if the caller passes a shorter array it will simply refuse to compile, which is likely what we want here

**Review by prrace — APPROVED** [review]
OK .. just do what you have.

**Comment by TheShermanTanker** [comment]
Thanks for the review. But I'd like to address your other concerns too if I can

**Review by aivanov-jdk — APPROVED** [review]
Looks good to me.

I added a suggestion to use the `enum` constant declared in `AwtToolkit` instead of hard-coding the size of the array.

File: src/java.desktop/windows/native/libawt/windows/awt_Component.cpp
Comment (TheShermanTanker) [reply src/java.desktop/windows/native/libawt/windows/awt_Component.cpp:3366]: > Will this resolve Phil's concern? Both arrays will use the same size.

@prrace Does this address your concerns? If it does I will commit it

Comment (prrace) [reply src/java.desktop/windows/native/libawt/windows/awt_Component.cpp:3366]: That's fine since that  what I was asking for in my comment a week ago
> Why can't we EVERYWHERE use the definition used to create the array passed in KB_STATE_SIZE = 256 instead of a mixture.

File: src/java.desktop/windows/native/libawt/windows/awt_Component.cpp
Comment (TheShermanTanker) [reply src/java.desktop/windows/native/libawt/windows/awt_Component.cpp:3368]: I can't quite comment on that since I don't really know what the purpose of the memmove is. What does @prrace think?

Comment (aivanov-jdk) [reply src/java.desktop/windows/native/libawt/windows/awt_Component.cpp]: Both `memcpy` and `memmove` do the same thing, namely each function copies bytes from one location into another location. The difference between these two functions is in whether buffers are allowed to overlap or not:

* [`memcpy`](https://en.cppreference.com/w/c/string/byte/memcpy): <q cite="https://en.cppreference.com/w/c/string/byte/memcpy">If the objects overlap […], the behavior is undefined.</q><br><q cite="https://en.cppreference.com/w/c/string/byte/memcpy#Notes"><em>`memcpy` is the fastest library routine for memory-to-memory copy.</em> It is usually more efficient than `strcpy`, which must scan the data it copies or `memmove`, which must take precautions to handle overlapping inputs.</q>
* [`memmove`](https://en.cppreference.com/w/c/string/byte/memmove): <q cite="https://en.cppreference.com/w/c/string/byte/memmove">The objects may overlap: copying takes place as if the characters were copied to a temporary character array and then the characters were copied from the array to `dest`.</q>

> we know for sure that `tmpState` and `kstate` do not overlap.

For this reason, `memcpy` is safe to use, and it is more efficient than `memmove`.

To be clear, I'm not proposing to incorporate this change under this bug. I just pointed out a possible optimisation.

**Review by tstuefe — APPROVED** [review]
Weird syntax to someone like me, but I can see the benefit.

Using memmove is so uncommon that it is usually a clear indication for a deliberate choice.

**Review by aivanov-jdk — APPROVED** [review]
Still looks good, except for the minor formatting comments.

**Comment by aivanov-jdk** [comment]
> Using memmove is so uncommon that it is usually a clear indication for a deliberate choice.

It may still be an accidental choice. I didn't find any code review for [JDK-6680988](https://bugs.openjdk.org/browse/JDK-6680988) where this line had been added.

**Review by prrace — APPROVED** [review]
I'd leave the memmove as it is unless someone can show a measurable benefit to changing to memcpy.

I do agree with the suggested formatting changes.

File: src/java.desktop/windows/native/libawt/windows/awt_Component.cpp
Comment (TheShermanTanker) [reply src/java.desktop/windows/native/libawt/windows/awt_Component.cpp:3365]: Will do

File: src/java.desktop/windows/native/libawt/windows/awt_Component.cpp
Comment (TheShermanTanker) [reply src/java.desktop/windows/native/libawt/windows/awt_Component.cpp:3368]: Ah, seems awt follows a different convention from HotSpot, which I'm nmore used to. I'll revert the formatting

**Comment by TheShermanTanker** [comment]
Sorry, am awaiting re\-review from all of you

**Review by tstuefe — APPROVED** [review]
Okay.

(Please don't go now and change all occurrences of `foo(pointer, size)` to something like `foo( [&] [size] )` :-) I dislike the increased cognitive load of this form, and prefer the classic pointer+size combo )

**Comment by TheShermanTanker** [comment]
> Okay.
> 
> (Please don't go now and change all occurrences of `foo(pointer, size)` to something like `foo( [&] [size] )` :-) I dislike the increased cognitive load of this form, and prefer the classic pointer+size combo )

Even if I wanted to, that wouldn't be possible, since this syntax is only allowed when passing in an array of known size directly to the method, and the array has to match the parameter declaration in the method exactly ;)

Will wait for approval from Phil and Alexey before pushing

**Review by aivanov-jdk — APPROVED** [review]

File: src/java.desktop/windows/native/libawt/windows/awt_Component.cpp
Comment (aivanov-jdk) [reply src/java.desktop/windows/native/libawt/windows/awt_Component.cpp]: I just searched for `sizeof` in the file. The next function has
```c
    memset(kbdState, 0, sizeof (kbdState));
```
with a space. Other usages in the file don't. I'm following the convention used in the file.

Please, don't change this single instance above, let it stay.

File: src/java.desktop/windows/native/libawt/windows/awt_Component.cpp
Comment (aivanov-jdk) [reply src/java.desktop/windows/native/libawt/windows/awt_Component.cpp]: Here, I'd remove the space after the opening parenthesis: `resetKbdState(BYTE`.

It's not worth changing at this point, though…

**Pull request closed without merging**
